### PR TITLE
Allow projections on sub-attributes

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Support projections on sub-attributes, e.g. a.b.c.
Previously, projections were only supported on top-level attributes. For example, in the query
```
FOR doc IN collection
  RETURN doc.a.b
```
the projection was just `a`.
With this change, the projection will be `a.b`.

In addition, an index can be used now to extract the projection data even for indexes on sub-attributes. 
If for the above query an index on `a.b` exists, it will be used now. Before this change, only an index on `a` could be used, but none on `a.b`.

In addition, this change now allows projections to be fed by an attribute from a combined index. For example, in the query
```
FOR doc IN collection
  RETURN doc.b
```
previously there has to exist on index on `b`. If an index on `[a, b]` existed, it wasn't used.
With this change, an index on `b` or an index on `[a, b]` can be used to feed to projection, thus removing the new to peek at the full documents.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

#### Related Information

- [x] There is a *JIRA Ticket number*: DEVSUP-624 

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

### CODE REVIEW

File: arangod/Aql/Ast.h [hunk 1]
```diff
@@ -402,14 +403,22 @@ class Ast {
   /// @brief count how many times a variable is referenced in an expression
   static size_t countReferences(AstNode const*, Variable const*);
 
-  /// @brief determines the top-level attributes in an expression, grouped by
+  /// @brief determines the top-level attributesused  in an expression, grouped by
```
Comment (jsteemann) [inline arangod/Aql/Ast.h:406]:
```suggestion
  /// @brief determines the top-level attributes used  in an expression, grouped by
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11450/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11459/

File: CHANGELOG [hunk 2]
```diff
@@ -1,6 +1,31 @@
 devel
 -----
 
+* Support projections on sub-attributes (`a.b.c`).
```
Comment (jsteemann) [inline CHANGELOG:4]:
```suggestion
* Support projections on sub-attributes (e.g. `a.b.c`).
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11483/

File: arangod/Aql/RegisterPlan.cpp [hunk 3]
```diff
@@ -450,9 +450,9 @@ void RegisterPlanT<T>::toVelocyPackEmpty(VPackBuilder& builder) {
   { VPackArrayBuilder guard(&builder); }
   builder.add(VPackValue("nrRegs"));
   { VPackArrayBuilder guard(&builder); }
-  // nrRegsHere is not used anymore and is intentionally left empty
-  // can be removed in ArangoDB 3.8
-  builder.add(VPackValue("nrRegsHere"));
+  // nrRegsHere is not used anymore and is intentionally left empty.
+  // it is only here for historical reasons.
+  // builder.add(VPackValue("nrRegsHere"));
```
Comment (danielhlarkin) [inline arangod/Aql/RegisterPlan.cpp:455]:
```suggestion

```

File: arangod/Aql/RemoteExecutor.cpp [hunk 4]
```diff
@@ -289,10 +289,12 @@ std::pair<ExecutionState, Result> ExecutionBlockImpl<RemoteExecutor>::initialize
   VPackBuilder builder(buffer, &options);
   builder.openObject(/*unindexed*/ true);
 
-  // Required for 3.5.* and earlier, dropped in 3.6.0
-  builder.add("exhausted", VPackValue(false));
-  // Used in 3.4.0 onwards
+  // Required for 3.5.* and earlier, dropped in 3.6.0.
+  // Just here for historical reasons:
+  //   builder.add("exhausted", VPackValue(false));
```
Comment (danielhlarkin) [inline arangod/Aql/RemoteExecutor.cpp:294]:
```suggestion
```

File: arangod/Aql/RegisterPlan.h [hunk 5]
```diff
@@ -151,7 +151,6 @@ struct RegisterPlanT final : public std::enable_shared_from_this<RegisterPlanT<T
 
   auto variableToRegisterId(Variable const* variable) const -> RegisterId;
 
-  // compatibility function for 3.6. can be removed in 3.8
```
Comment (danielhlarkin) [inline arangod/Aql/RegisterPlan.h:154]: Is the comment inaccurate, or should the function be removed?

Comment (jsteemann) [reply arangod/Aql/RegisterPlan.h:154]: Unfortunately, the comment was inaccurate :smile:

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11485/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11486/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11494/

File: arangod/Aql/AttributeNamePath.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "Aql/AttributeNamePath.h"
+#include "Basics/StaticStrings.h"
+#include "Basics/debugging.h"
+#include "Basics/fasthash.h"
+
+#include <algorithm>
+
+#include <boost/functional/hash.hpp>
+
```
Comment (jsteemann) [inline arangod/Aql/AttributeNamePath.cpp:32]:
```suggestion
```

File: arangod/Aql/Ast.h [hunk 7]
```diff
@@ -402,14 +403,22 @@ class Ast {
   /// @brief count how many times a variable is referenced in an expression
   static size_t countReferences(AstNode const*, Variable const*);
 
-  /// @brief determines the top-level attributes in an expression, grouped by
+  /// @brief determines the top-level attributes used  in an expression, grouped by
```
Comment (jsteemann) [inline arangod/Aql/Ast.h:406]:
```suggestion
  /// @brief determines the top-level attributes used in an expression, grouped by
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11512/

**Comment by jsteemann** [comment]
Tests blue

File: arangod/Aql/AttributeNamePath.h [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+#include <vector>
+
+namespace arangodb {
+namespace aql {
+
+/// @brief helper class to handle top level and nested attribute paths.
+/// a top-level attribute (e.g. _key) will be stored in a vector with a single
+/// element. nested attributes (e.g. a.b.c) will be stored in a vector with
+/// multiple elements
+struct AttributeNamePath {
```
Comment (graetzer) [inline arangod/Aql/AttributeNamePath.h:38]: didn't we have something like this for index fields ?

Comment (jsteemann) [reply arangod/Aql/AttributeNamePath.h:38]: Yes, we have `AttributeName` in lib/Basics/AttributeNameParser.h, but that is slightly different. It supports attribute name expansion (e.g. `a.b[*].c`), which means one cannot simply created a vector of strings but only of {string, bool} pairs. This makes it really annoying and more complicated to use in cases when just non-expanded attribute names are needed.

I can fuse the two classes, making things a bit more inefficient and harder to use, for the benefit of having less code around. Not sure what is better here.

**Review by graetzer — APPROVED** [review]

**Pull request merged**